common/src/main/java/com/microsoft/identity/common/internal/controllers/TokenCommand.java

```diff
@@ -76,28 +76,41 @@ public TokenCommand(@NonNull final Context context,
     }
 
     @Override
-    public AcquireTokenResult execute() throws InterruptedException, ExecutionException, IOException, ClientException, UiRequiredException, ArgumentException, ServiceException {
+    public AcquireTokenResult execute()
+            throws InterruptedException, ExecutionException, IOException, ClientException,
+            UiRequiredException, ArgumentException, ServiceException {
         AcquireTokenResult result = null;
         final String methodName = ":execute";
 
-        for(BaseController controller : mControllers) {
+        for (int ii = 0; ii < mControllers.size(); ii++) {
+            final BaseController controller = mControllers.get(ii);
+
             try {
                 com.microsoft.identity.common.internal.logging.Logger.verbose(
                         TAG + methodName,
-                        "Executing with controller: " + controller.getClass().getSimpleName()
+                        "Executing with controller: "
+                                + controller.getClass().getSimpleName()
+                );
+
+                result = controller.acquireTokenSilent(
+                        (AcquireTokenSilentOperationParameters) getParameters()
                 );
-                result = controller.acquireTokenSilent((AcquireTokenSilentOperationParameters) getParameters());
-                if(result.getSucceeded()){
+
+                if (result.getSucceeded()) {
                     com.microsoft.identity.common.internal.logging.Logger.verbose(
                             TAG + methodName,
-                            "Executing with controller: " + controller.getClass().getSimpleName() + ": Succeeded"
+                            "Executing with controller: "
+                                    + controller.getClass().getSimpleName()
+                                    + ": Succeeded"
                     );
+
                     return result;
                 }
-            }catch(UiRequiredException e){
-                if(e.getErrorCode().equals(UiRequiredException.INVALID_GRANT)){
+            } catch (UiRequiredException e) {
+                if (e.getErrorCode().equals(UiRequiredException.INVALID_GRANT) // was invalid_grant
+                        && mControllers.size() > ii + 1) { // isn't the last controller we can try
```

Comment:
👍

Reply:
Resolves https://github.com/AzureAD/azure-activedirectory-library-for-android/issues/1394